### PR TITLE
quic: Fix QuicFilterManagerConnectionImpl::ConnectionInfoProviderSharedPtr

### PR DESCRIPTION
**Pull request by RyanTheOptimist**
quic: Fix QuicFilterManagerConnectionImpl::ConnectionInfoProviderSharedPtr

If QuicFilterManagerConnectionImpl::ConnectionInfoProviderSharedPtr is
called after the connection is closed, return nullptr.

Risk Level: Low
Testing: New unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: source/common/quic/quic_filter_manager_connection_impl.h [hunk 1]
```diff
@@ -64,6 +64,9 @@ class QuicFilterManagerConnectionImpl : public Network::ConnectionImplBase,
     return network_connection_->connectionSocket()->connectionInfoProvider();
```
Comment (danzh2010) [inline source/common/quic/quic_filter_manager_connection_impl.h:64]: Does this one need the same fix as well?

File: source/common/quic/quic_filter_manager_connection_impl.h [hunk 2]
```diff
@@ -64,6 +64,9 @@ class QuicFilterManagerConnectionImpl : public Network::ConnectionImplBase,
...
   }
   Network::ConnectionInfoProviderSharedPtr connectionInfoProviderSharedPtr() const override {
+    if (!network_connection_ || !network_connection_->connectionSocket()) {
```
Comment (danzh2010) [inline source/common/quic/quic_filter_manager_connection_impl.h:67]: Does this break existing tests?

Comment (RyanTheOptimist) [reply source/common/quic/quic_filter_manager_connection_impl.h:67]: I hope not :) If either of those are null when this method is called, we would see a crash in the old code. With the new code, we don't crash, we just return null. CI is green, so I think we're ok.

File: source/common/quic/quic_filter_manager_connection_impl.h (hunk 1 above)
Comment (RyanTheOptimist) [reply source/common/quic/quic_filter_manager_connection_impl.h:64]: I don't think this one is fixable, at least not without a larger refactor of this class (which maybe we should do?). This method returns a reference to a ConnectionInfoSetter. But if the network_connection_ or the network_connection_->connectionSocket() are null then this method will return a null reference (which is illegal). So presumably this method must only be called when they are non-null? If that is not a valid assumption then perhaps we need to refactor this class to make this information always available?

Comment (danzh2010) [reply source/common/quic/quic_filter_manager_connection_impl.h:64]: Good point! Can you add ENVOY_BUG before returning?

Comment (RyanTheOptimist) [reply source/common/quic/quic_filter_manager_connection_impl.h:66]: Done!

**Review by danzh2010 — APPROVED** [review]
Thanks for fixing this!

**Review by alyssawilk — APPROVED** [review]

File: test/common/quic/quic_filter_manager_connection_impl_test.cc [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+  void closeNow() {
+    const quic::QuicConnectionCloseFrame frame;
+    quic::ConnectionCloseSource source = quic::ConnectionCloseSource::FROM_SELF;
+    const quic::ParsedQuicVersion version = quic::AllSupportedVersions().front();
+    onConnectionCloseEvent(frame, source, version);
+  }
+};
+
+TEST(QuicFilterManagerConnectionImplTest, ConnectionInfoProviderSharedPtr) {
```
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply test/common/quic/quic_filter_manager_connection_impl_test.cc]: Heh, I did consider it initially and wasn't sure if it'd be worth doing. But if it also occurred to you then I'm sure it is. Done!

**Review by RyanTheOptimist — COMMENTED** [review]
Thanks

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**